pkg/scheduler/preemption/preemption_test.go

```diff
@@ -0,0 +1,587 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package preemption
+
+import (
+	"context"
+	"sort"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/go-logr/logr/testr"
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client/fake"
+
+	kueue "sigs.k8s.io/kueue/apis/kueue/v1alpha2"
+	"sigs.k8s.io/kueue/pkg/cache"
+	"sigs.k8s.io/kueue/pkg/constants"
+	"sigs.k8s.io/kueue/pkg/scheduler/flavorassigner"
+	utiltesting "sigs.k8s.io/kueue/pkg/util/testing"
+	"sigs.k8s.io/kueue/pkg/workload"
+)
+
+func TestPreemption(t *testing.T) {
+	flavors := []*kueue.ResourceFlavor{
+		utiltesting.MakeResourceFlavor("default").Obj(),
+		utiltesting.MakeResourceFlavor("alpha").Obj(),
+		utiltesting.MakeResourceFlavor("beta").Obj(),
+	}
+	clusterQueues := []*kueue.ClusterQueue{
+		utiltesting.MakeClusterQueue("standalone").
+			Resource(utiltesting.MakeResource(corev1.ResourceCPU).
+				Flavor(utiltesting.MakeFlavor("default", "6").Obj()).
+				Obj()).
+			Resource(utiltesting.MakeResource(corev1.ResourceMemory).
+				Flavor(utiltesting.MakeFlavor("alpha", "3Gi").Obj()).
+				Flavor(utiltesting.MakeFlavor("beta", "3Gi").Obj()).
+				Obj()).
+			Preemption(kueue.ClusterQueuePreemption{
+				WithinClusterQueue: kueue.PreemptionPolicyLowerPriority,
+			}).
+			Obj(),
+		utiltesting.MakeClusterQueue("c1").
+			Cohort("cohort").
+			Resource(utiltesting.MakeResource(corev1.ResourceCPU).
+				Flavor(utiltesting.MakeFlavor("default", "6").Obj()).
+				Obj()).
+			Preemption(kueue.ClusterQueuePreemption{
+				WithinClusterQueue: kueue.PreemptionPolicyLowerPriority,
+				WithinCohort:       kueue.PreemptionPolicyReclaimFromLowerPriority,
+			}).
+			Obj(),
+		utiltesting.MakeClusterQueue("c2").
+			Cohort("cohort").
+			Resource(utiltesting.MakeResource(corev1.ResourceCPU).
+				Flavor(utiltesting.MakeFlavor("default", "6").Obj()).
+				Obj()).
+			Preemption(kueue.ClusterQueuePreemption{
+				WithinClusterQueue: kueue.PreemptionPolicyNever,
+				WithinCohort:       kueue.PreemptionPolicyReclaimFromAny,
+			}).
+			Obj(),
+	}
+	cases := map[string]struct {
+		admitted      []kueue.Workload
+		incoming      *kueue.Workload
+		targetCQ      string
+		assignment    flavorassigner.Assignment
+		wantPreempted sets.Set[string]
+	}{
```

Comment:
A suggestion for a weird test: Is it worth it to have a test where a workload has two podSets, each one require preemption for the same resource, but they get different flavors after preemption?

Reply:
Oh, I missed this comment. That's a good test case, I'll add one.

Reply:
Added